HDDS-11298. Support owner field in the S3 listBuckets request

hevinhsu opened the pull request:
## What changes were proposed in this pull request?
Add owner field in the S3 listBuckets request.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11298

## How was this patch tested?
https://github.com/hevinhsu/ozone/actions/runs/14570769206

jojochuang commented:
@hevinhsu thanks for opening a PR. Do you have an Apache jira id? Id' like to add you as the assignee. Thanks.

jojochuang commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ListBucketResponse.java:
```diff
@@ -39,6 +39,9 @@ public class ListBucketResponse {
   private List<BucketMetadata> buckets = new ArrayList<>();
 
+  @XmlElement(name = "Owner")
+  private S3Owner owner;
```
Unrelated, but S3Owner javadoc is not useful. We should fix it in another task.

    /**
     * TODO: javadoc.
     */
    @XmlAccessorType(XmlAccessType.FIELD)
    @XmlRootElement(name = "Owner")
    public class S3Owner {

jojochuang commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/RootEndpoint.java:
```diff
@@ -61,6 +62,9 @@ public Response get()
       }
 
+      OzoneVolume volume = getVolume();
+      response.setOwner(new S3Owner(volume.getName(), volume.getName()));
```
this line doesn't look right. The owner is the user owner of the volume, not the name of the volume

jojochuang commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/RootEndpoint.java:
```diff
@@ -61,6 +62,9 @@ public Response get()
       }
 
+      OzoneVolume volume = getVolume();
+      response.setOwner(new S3Owner(volume.getName(), volume.getName()));
```
```suggestion
      response.setOwner(new S3Owner(0, volume.getOwner()));
```

The first argument of S3Owner is ID, which is a 64-character encoded unique user name. We dont' support it in Ozone. https://docs.aws.amazon.com/accounts/latest/reference/manage-acct-identifiers.html

@ivandika3 any ideas about the owner id?

jojochuang replied:
I think we can make it simple. See how MinIO does it: https://github.com/minio/minio/issues/4553
Ceph S3 does not compute a user name hash either.

ivandika3 reviewed (commented):
Thanks @hevinhsu for the patch.

Overall comments
- Please help to add S3 SDK integration tests for the listBuckets owner behavior (see AbstractS3SDKV1Tests and AbstractS3SDKV2Tests)
- Also please add the acceptance test in bucketlist.robot

To discuss (cc: @jojochuang , @xichen01)
- The current situation is that currently of listBuckets will list all the buckets under the /s3v volume, whereas the ListBuckets API specification will only list all the buckets owned by the user. If we want to follow this, then the of listBuckets flow needs to be changed to filter our all the buckets that the user does not own, this might cause regressions on some users since the buckets that were returned before are now excluded (most users will not be able to list the buckets at all).
- If multi tenancy is disabled, we will use the default /s3v volume, which is initialized in OM using `OzoneManager#addS3GVolumeToDB`, the owner of this volume will be the UGI that starts the OzoneManager for the first time. Meaning that the S3 owner returned will always be the same. This is not compatible with the S3 API spec. Is this ok?

jojochuang commented:
> The current situation is that currently of listBuckets will list all the buckets under the /s3v volume, whereas the ListBuckets API specification will only list all the buckets owned by the user. If we want to follow this, then the of listBuckets flow needs to be changed to filter our all the buckets that the user does not own, this might cause regressions on some users since the buckets that were returned before are now excluded (most users will not be able to list the buckets at all).

I think this needs some clarification.

If ACL is enabled, it should forbid users from listing all buckets: https://github.com/apache/ozone/blob/42133076373dd3c1b72f0f8028771166a0594b50/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L2908-L2912

jojochuang commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/RootEndpoint.java:
```diff
@@ -61,6 +62,9 @@ public Response get()
         throw e;
       }
 
+      OzoneVolume volume = getVolume();
```
getVolume() is an RPC to Ozone Manager, and listS3Buckets(null) (above) internally invokes getVolume().
So this is redundant and put extra latency/overhead at OM. Can we refactor the code so getVolume() is called only once.

hevinhsu replied:
@jojochuang  Thanks for the feedback!

I'm currently using a functional interface to inject the owner into the response. This allows me to reuse the ```getVolume()``` error handling logic inside BaseEndpoint and avoid passing the response object into it.

The goal is to keep error handling consistent and isolate response construction from the base layer.

Could you help confirm if this approach aligns with our design expectations?

hevinhsu commented on hadoop-ozone/integration-test-s3/src/test/java/org/apache/hadoop/ozone/s3/S3ClientFactory.java:
```diff
@@ -158,4 +160,54 @@ public S3Client createS3ClientV2(boolean enablePathStyle) throws Exception {
+   * @return S3AsyncClient
+   * @throws URISyntaxException if there is an error creating the client
+   */
+  public S3AsyncClient createS3AsyncClientV2() throws Exception {
```
Hi,
I’ve implemented the S3AsyncClient based on the TODO in this file: [link](https://github.com/apache/ozone/blob/master/hadoop-ozone/integration-test-s3/src/test/java/org/apache/hadoop/ozone/s3/awssdk/v2/AbstractS3SDKV2Tests.java#L93-L97)

I haven’t worked on the crt-based-s3-client test yet, since it would require introducing a new dependency.

Just wanted to confirm—are both of these tests expected to be included as part of this issue?

Thanks!

ivandika3 replied:
@hevinhsu Thanks for considering `S3AsyncClient`. I think we can introduce `S3AsyncClient` in another task which will add the S3 Async client in all existing tests in  `AbstractS3SDKV2Tests`. We can exclude `S3AsyncClient` and  focus on the synchronous `S3Client` for this patch.

hevinhsu replied:
Thanks for the reviewe.
I'll remove ```S3AsyncClient``` in this PR.

ivandika3 commented on hadoop-ozone/integration-test-s3/src/test/java/org/apache/hadoop/ozone/s3/awssdk/v2/AbstractS3SDKV2Tests.java:
```diff
@@ -152,116 +133,6 @@ public void testPutObject() {
     assertEquals("\"37b51d194a7513e45b56f6524f2d51f2\"", getObjectResponse.eTag());
   }
 
-  @Test
-  public void testListObjectsMany() throws Exception {
-    testListObjectsMany(false);
-  }
-
-  @Test
-  public void testListObjectsManyV2() throws Exception {
-    testListObjectsMany(true);
-  }
```
Seems there is a mistake when merging to master and all of the `AbstractS3SDKV2Tests` changes in https://github.com/apache/ozone/pull/8356 were removed. Please kindly revert all the deletions in this file.

hevinhsu replied:
Oops, sorry about that. I'll fix it soon.

ivandika3 commented on hadoop-ozone/integration-test-s3/src/test/java/org/apache/hadoop/ozone/s3/awssdk/v1/AbstractS3SDKV1Tests.java:
```diff
@@ -249,7 +250,7 @@ public void testBucketACLOperations() {
 
   @Test
-  public void testListBuckets() {
+  public void testListBuckets() throws IOException {
```
We can add this `testListBuckets` to `AbstractS3SDKV2Tests`.

hevinhsu commented:
Thanks everyone for the thoughtful feedback!
 
I've:

- Added acceptance test
- Added integration tests for both S3 SDK v1 and v2
- Fixed incorrect merge code
 
@jojochuang @ivandika3 

Feel free to take another look when you get a chance — let me know if there’s anything else I should address.

hevinhsu commented:
@ivandika3 Thanks for the review and the suggestion!

I've opened the Jira ticket here: https://issues.apache.org/jira/browse/HDDS-13032
Let me know if you think any part of it needs adjustment.

ivandika3 commented:
@hevinhsu Thanks for creating a ticket

Could you help to do one last merge to master? Afterwards, I'll start the CI.

hevinhsu commented:
@ivandika3 Thanks a lot! I've merged the latest changes from master.

ivandika3 commented:
Thank you @hevinhsu for the patch and @jojochuang for the review.
